src/PolynomialFunction.java

```diff
@@ -3,15 +3,18 @@
 import java.util.*;
 
 public class PolynomialFunction implements ParametricFunction {
-
-    private double[] coefficients;
     public String functionString;
+    private double[] coefficients;
+    public List<Double> roots;
+    public List<Double> extremePoints;
     private static final Pattern TERM_PATTERN = Pattern.compile("([-+]?\\s*\\d*\\.?\\d*(?:/\\d+)*)?x(\\^(-?\\d+))?|([-+]?\\s*\\d+(/\\d+)?)");
```

Comment:
again, these can be calculated and cached wherever needed. No need for state here, just have 'calcRoots' and 'getExtremePoints' return the lists directly. Also, if the TERM_PATTERN is only required in one place, it might be more sensible to declare it there.

Reply:
The reason for that was that these values are only calculated once during creation of the polynomial instead of everytime the screen gets repainted => saving performance
But please correct me if i see something wrong here.
As for the TERM Pattern you are completely right. It was originally placed there for performance but i neglected the fact that the parsing only happens once for each function

Reply:
I think the GraphPanel may be the better location for this

Reply:
Done